### PR TITLE
Replace ON DUPLICATE KEY UPDATE with new Syntax ON CONFLICT

### LINKED ISSUE #10: [CrateDB] Replace ON DUPLICATE KEY UPDATE with new Syntax ON CONFLICT
Replace CrateDB SQL Statements which use `ON DUPLICATE KEY UPDATE` with `INSERT INTO ON CONFLICT` https://crate.io/docs/crate/reference/en/latest/sql/statements/insert.html#on-conflict-do-update-set

New Syntax Example
````
INSERT INTO statusengine_servicestatus
    (hostname, service_description, status_update_time, output, long_output, perfdata, current_state, current_check_attempt, max_check_attempts, last_check, next_check, is_passive_check, last_state_change, last_hard_state_change, last_hard_state, is_hardstate, last_notification, next_notification, notifications_enabled, problem_has_been_acknowledged, acknowledgement_type, passive_checks_enabled, active_checks_enabled, event_handler_enabled, flap_detection_enabled, is_flapping, latency, execution_time, scheduled_downtime_depth, process_performance_data, obsess_over_service, normal_check_interval, retry_check_interval, check_timeperiod, node_name, last_time_ok, last_time_warning, last_time_critical, last_time_unknown, current_notification_number, percent_state_change, event_handler, check_command)
    VALUES('a47af17f-15d9-41af-ba9d-cb89c8d88e18','c3b06d13-cb3b-47d1-9342-33a4e41bc7d0',1537949703,'OK - load average: 0.63, 0.98, 1.07','','load1=0.630;20.000;35.000;0; load5=0.980;25.000;28.000;0; load15=1.070;25.000;28.000;0;',0,1,3,1537949703,1537949763,true,1537172439,1535117522,0,true,0,0,true,false,0,true,true,true,false,false,0.198347,0.01918,0,true,true,60,60,'41012866-6114-4853-9caf-6ffd19954e50','staging-nightly',1537949703,1535117522,1535117462,1537172439,0,0,'','some we replace is this auto updatE?'), ('b383f0b7-acfa-4337-9456-b3220b876ed6','94c964e2-5256-4977-bd7b-fbed81e764d3',1537949703,'UNKNOWN: Unknown - woher soll ich denn das wissen','','',3,3,3,1537949703,1537950603,true,1524728151,1524728151,3,true,0,1537949703,true,false,0,true,true,true,false,false,0.217184,0.002269,0,false,true,900,60,'41012866-6114-4853-9caf-6ffd19954e50','staging-nightly',0,0,0,1537949703,0,0,'','d4f3524b-d90f-4562-a2fb-c70d823f66ef!3!Unknown - woher soll ich denn das wissen'), ('0cf3730c-4e57-401d-9fa9-c9f698c4e405','f24a6407-3c39-4d1a-af93-a8e9fadc97ef',1537949703,'Remote command execution failed: @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@','','',3,3,3,1537949643,1537949763,true,1521197992,1521197992,3,true,0,1537949643,true,false,0,true,true,true,false,false,0.302169,0.040419,0,true,true,60,60,'41012866-6114-4853-9caf-6ffd19954e50','staging-nightly',1520682803,0,1521197992,1537949643,0,0,'','fd52f68f-e0f9-4c69-9f69-bc1b2d97188a!monitoring!sudo /usr/lib/nagios/plugins/check_zfs -w 80 -c 90 -p rpool'), ('9a456851-57ca-4770-b335-bd9276183642','9d56148a-7067-4566-b5bc-d4b53bbca297',1537949703,'OK: Alles OK','','',0,1,3,1537948803,1537950603,true,1506687048,1506687048,0,true,0,0,true,false,0,true,true,true,false,false,0.328132,0.001923,0,false,true,900,60,'41012866-6114-4853-9caf-6ffd19954e50','staging-nightly',1537948803,0,0,1506687048,0,0,'','d4f3524b-d90f-4562-a2fb-c70d823f66ef!0!Alles OK'), ('9a456851-57ca-4770-b335-bd9276183642','9d56148a-7067-4566-b5bc-d4b53bbca297',1537949703,'OK: Alles OK','','',0,1,3,1537949703,1537950603,true,1506687048,1506687048,0,true,0,0,true,false,0,true,true,true,false,false,0.327552,0.001741,0,false,true,900,60,'41012866-6114-4853-9caf-6ffd19954e50','staging-nightly',1537949703,0,0,1506687048,0,0,'','d4f3524b-d90f-4562-a2fb-c70d823f66ef!0!Alles OK')
    ON CONFLICT (hostname, service_description) DO UPDATE SET status_update_time = excluded.status_update_time, output = excluded.output, long_output = excluded.long_output, perfdata = excluded.perfdata, current_state = excluded.current_state, current_check_attempt = excluded.current_check_attempt, max_check_attempts = excluded.max_check_attempts, last_check = excluded.last_check, next_check = excluded.next_check, is_passive_check = excluded.is_passive_check, last_state_change = excluded.last_state_change, last_hard_state_change = excluded.last_hard_state_change, last_hard_state = excluded.last_hard_state, is_hardstate = excluded.is_hardstate, last_notification = excluded.last_notification, next_notification = excluded.next_notification, notifications_enabled = excluded.notifications_enabled, problem_has_been_acknowledged = excluded.problem_has_been_acknowledged, acknowledgement_type = excluded.acknowledgement_type, passive_checks_enabled = excluded.passive_checks_enabled, active_checks_enabled = excluded.active_checks_enabled, event_handler_enabled = excluded.event_handler_enabled, flap_detection_enabled = excluded.flap_detection_enabled, is_flapping = excluded.is_flapping, latency = excluded.latency, execution_time = excluded.execution_time, scheduled_downtime_depth = excluded.scheduled_downtime_depth, process_performance_data = excluded.process_performance_data, obsess_over_service = excluded.obsess_over_service, normal_check_interval = excluded.normal_check_interval, retry_check_interval = excluded.retry_check_interval, check_timeperiod = excluded.check_timeperiod, node_name = excluded.node_name, last_time_ok = excluded.last_time_ok, last_time_warning = excluded.last_time_warning, last_time_critical = excluded.last_time_critical, last_time_unknown = excluded.last_time_unknown, current_notification_number = excluded.current_notification_number, percent_state_change = excluded.percent_state_change, event_handler = excluded.event_handler, check_command = excluded.check_command ;
````

Old Syntax
````SQL
Short Example
INSERT INTO table (col1, col2, col3) VALUES(?,?, ?)
ON DUPLICATE KEY UPDATE
col2=VALUES(col2), col3=VALUES(col3);
````
New one
````SQL
INSERT INTO table (col1, col2, col3) VALUES(?,?,?)
ON CONFLICT (col1)
DO UPDATE SET col2 = excluded.col2, col3 = excluded.col3;
````

### PR DESCRIPTION
**Pull request by vanyasem**
Closes #10

### CODE REVIEW

**Comment by nook24** [comment]
Thanks @vanyasem for your contribution.
Unfortunately it's not done by a simple string replace, because also the primary key needs to be specified within the new syntax.

Old syntax
````SQL
INSERT INTO table (col1, col2, col3) VALUES(?,?, ?)
ON DUPLICATE KEY UPDATE
col2=VALUES(col2), col3=VALUES(col3);
````

New syntax
````SQL
INSERT INTO table (col1, col2, col3) VALUES(?,?,?)
ON CONFLICT (col1)
DO UPDATE SET col2 = excluded.col2, col3 = excluded.col3;
````
In addition you also replace the MySQL statements, which does not support `ON CONFLIC DO` afaik.

**Comment by vanyasem** [comment]
Isn't the primary key optional in the new syntax? @nook24

**Comment by vanyasem** [comment]
And with MySQL - you're correct, that's my bad

**Comment by vanyasem** [comment]
It would be great if you re-opened the PR so that I can push fixes to it. Why would you close it in the first place?

**Comment by nook24** [comment]
Of cource, just an accident...

If i try to execute the new query without passing the PK I run into an `SQLActionException[SQLParseException: line 4:20: mismatched input 'UPDATE' expecting 'NOTHING']`

Schema is defined in this file: https://github.com/statusengine/worker/blob/master/lib/crateDB.sql

**Comment by vanyasem** [comment]
Should be good now

**Comment by nook24** [comment]
Good to see you again 🙂.

Unfortunately there is still a little issue with the syntax.  After conflict the columns needs to be passed as  col = excluded.col instead of col=VALUES(col). 

Send from mobile

**Comment by vanyasem** [comment]
Should those https://github.com/statusengine/worker/pull/12/files#diff-92f70bc07cdac9cee023e69579b196d7L157 also be updated, or should they remain the same? @nook24

**Comment by vanyasem** [comment]
I used sed, and it seems that it also affected those entries, not sure if that's correct

File: src/Backends/Crate/SqlObjects/CrateServiceScheduleddowntime.php [hunk 1]
```diff
@@ -139,7 +139,7 @@ public function getQuery(Downtime $Downtime) {
         if ($Downtime->wasDowntimeAdded() || $Downtime->wasRestoredFromRetentionDat() || $Downtime->wasDowntimeStarted()) {
             $dynamicFields = [
                 'insert' => ['was_started', 'actual_start_time'],
-                'update' => ['was_started=VALUES(was_started)', 'actual_start_time=VALUES(actual_start_time)']
+                'update' => ['was_started = excluded.was_started)', 'actual_start_time = excluded.actual_start_time)']
```
Comment (nook24) [inline src/Backends/Crate/SqlObjects/CrateServiceScheduleddowntime.php]: Closing bracket ) needs to be removed

File: src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php [hunk 2]
```diff
@@ -166,15 +166,15 @@ private function getQuery(Downtime $Downtime) {
...
+                'update' => ['was_started = excluded.was_started)', 'actual_start_time = excluded.actual_start_time)']
             ];
             return $this->buildQueryString($dynamicFields);
         }
 
         if ($Downtime->wasDowntimeStopped() || $Downtime->wasDowntimeDeleted()) {
             $dynamicFields = [
                 'insert' => ['actual_end_time', 'was_cancelled'],
-                'update' => ['actual_end_time=VALUES(actual_end_time)', 'was_cancelled=VALUES(was_cancelled)']
+                'update' => ['actual_end_time = excluded.actual_end_time)', 'was_cancelled = excluded.was_cancelled)']
```
Comment (nook24) [inline src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php]: Closing bracket ) needs to be removed

File: src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php [hunk 3]
```diff
@@ -166,15 +166,15 @@ private function getQuery(Downtime $Downtime) {
         if ($Downtime->wasDowntimeStarted()) {
             $dynamicFields = [
                 'insert' => ['was_started', 'actual_start_time'],
-                'update' => ['was_started=VALUES(was_started)', 'actual_start_time=VALUES(actual_start_time)']
+                'update' => ['was_started = excluded.was_started)', 'actual_start_time = excluded.actual_start_time)']
```
Comment (nook24) [inline src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php]: Closing bracket ) needs to be removed

File: src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php [hunk 4]
```diff
@@ -154,10 +154,10 @@ private function getQuery(Downtime $Downtime) {
             $dynamicFields = [
                 'insert' => ['was_started', 'actual_start_time', 'actual_end_time', 'was_cancelled'],
                 'update' => [
-                    'was_started=VALUES(was_started)',
-                    'actual_start_time=VALUES(actual_start_time)',
-                    'actual_end_time=VALUES(actual_end_time)',
-                    'was_cancelled=VALUES(was_cancelled)'
+                    'was_started = excluded.was_started)',
```
Comment (nook24) [inline src/Backends/Crate/SqlObjects/CrateServiceDowntimehistory.php]: Closing bracket ) needs to be removed

File: src/Backends/Crate/Crate.php [hunk 5]
```diff
@@ -101,7 +101,7 @@ public function saveNodeName($nodeName = null, $startTime = null) {
 
         $this->connect();
         $query = $this->Connection->prepare('INSERT INTO statusengine_nodes (node_name, node_version, node_start_time)
-          VALUES(?,?,?) ON DUPLICATE KEY UPDATE node_version=VALUES(node_version), node_start_time=VALUES(node_start_time)');
+          VALUES(?,?,?) ON CONFLICT (node_name) DO UPDATE SET node_version = excluded.node_version, node_start_time = excluded.node_start_time');
```
Comment (nook24) [inline src/Backends/Crate/Crate.php]: Closing bracket ) needs to be removed

**Comment by vanyasem** [comment]
I manually went over all closing brackets, but I guess I missed some of them. Sorry

**Comment by nook24** [comment]
Good Job, thanks a lot!

**Pull request merged**